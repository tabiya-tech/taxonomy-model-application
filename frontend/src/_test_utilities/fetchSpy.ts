export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
