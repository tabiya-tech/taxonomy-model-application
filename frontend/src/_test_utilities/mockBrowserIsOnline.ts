import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
