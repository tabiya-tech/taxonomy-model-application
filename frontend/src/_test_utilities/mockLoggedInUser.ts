import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
