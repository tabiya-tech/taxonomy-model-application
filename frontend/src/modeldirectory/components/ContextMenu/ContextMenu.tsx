import * as React from "react";
import { ListItemIcon, ListItemText, Menu, MenuItem, Typography } from "@mui/material";
import CloudDownloadIcon from "@mui/icons-material/CloudDownload";

export interface ContextMenuProps {
  anchorEl?: HTMLElement | null;
  open: boolean;
  isExportDisabled: boolean;
  notifyOnClose: () => void;
  notifyOnExport: () => void;
}

const uniqueId = "715a535d-ca4b-4252-b967-8d5b5b2381b9";

export const DATA_TEST_ID = {
  CONTEXT_MENU: `context-menu-${uniqueId}`,
  MENU_ITEM_EXPORT: `menu-item-export-${uniqueId}`,
};

const ContextMenu = (props: Readonly<ContextMenuProps>) => {
  function handleExport() {
    props.notifyOnClose();
    props.notifyOnExport();
  }

  return (
    <Menu
      data-testid={DATA_TEST_ID.CONTEXT_MENU}
      elevation={2}
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "center",
      }}
      transformOrigin={{
        vertical: "top",
        horizontal: "right",
      }}
      anchorEl={props.anchorEl}
      open={props.open}
      onClose={props.notifyOnClose}
    >
      <MenuItem
        onClick={() => handleExport()}
        component="button"
        data-testid={DATA_TEST_ID.MENU_ITEM_EXPORT}
        disabled={props.isExportDisabled}
      >
        <ListItemIcon>
          <CloudDownloadIcon color="primary" />
        </ListItemIcon>
        <ListItemText>
          <Typography variant={"caption"} color={"primary"}>
            Export
          </Typography>
        </ListItemText>
      </MenuItem>
    </Menu>
  );
};

export default ContextMenu;
