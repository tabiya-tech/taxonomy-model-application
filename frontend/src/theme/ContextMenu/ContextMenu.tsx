import { Icon, ListItemIcon, ListItemText, Menu, MenuItem, Typography } from "@mui/material";
import { MenuItemConfig } from "./menuItemConfig.types";

export interface ContextMenuProps {
  items: MenuItemConfig[];
  open: boolean;
  anchorEl: HTMLElement | null;
  notifyOnClose: () => void;
}

const uniqueId = "b7499b01-8082-4209-8667-c7d559a70caf";
export const DATA_TEST_ID = {
  MENU: `${uniqueId}-menu`,
  MENU_ITEM: `${uniqueId}-menu-item`,
  MENU_ITEM_ICON: `${uniqueId}-menu-item-icon`,
  MENU_ITEM_TEXT: `${uniqueId}-menu-item-text`,
};

function ContextMenu(props: Readonly<ContextMenuProps>) {
  const handleItemClick = (item: MenuItemConfig) => {
    props.notifyOnClose();
    item.action();
  };

  return (
    <Menu
      elevation={2}
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "center",
      }}
      transformOrigin={{
        vertical: "top",
        horizontal: "right",
      }}
      anchorEl={props.anchorEl}
      open={props.open}
      onClose={props.notifyOnClose}
      data-testid={DATA_TEST_ID.MENU}
    >
      {props.items.map((item) => (
        <MenuItem
          onClick={() => handleItemClick(item)}
          data-testid={DATA_TEST_ID.MENU_ITEM}
          disabled={item.disabled}
          key={item.id}
        >
          {item.icon && (
            <ListItemIcon data-testid={DATA_TEST_ID.MENU_ITEM_ICON}>
              <Icon color="primary">{item.icon}</Icon>
            </ListItemIcon>
          )}
          <ListItemText data-testid={DATA_TEST_ID.MENU_ITEM_TEXT}>
            <Typography variant="caption" color="primary">
              {item.text}
            </Typography>
          </ListItemText>
        </MenuItem>
      ))}
    </Menu>
  );
}

export default ContextMenu;
