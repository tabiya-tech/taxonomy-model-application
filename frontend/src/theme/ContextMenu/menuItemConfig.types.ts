import { ReactNode } from "react";

export type MenuItemConfig = {
  id: string;
  text: string;
  icon?: ReactNode;
  disabled: boolean;
  action: () => void;
};
