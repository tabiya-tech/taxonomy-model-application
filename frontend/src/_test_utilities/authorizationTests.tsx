import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
