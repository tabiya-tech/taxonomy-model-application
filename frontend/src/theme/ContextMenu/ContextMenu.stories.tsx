import { Meta, StoryObj } from "@storybook/react";
import * as React from "react";
import { useEffect } from "react";
import Box from "@mui/material/Box";
import { action } from "@storybook/addon-actions";
import CloudDownloadIcon from "@mui/icons-material/CloudDownload";
import { SendOutlined } from "@mui/icons-material";
import ContextMenu, { ContextMenuProps } from "./ContextMenu";

const meta: Meta<typeof ContextMenu> = {
  title: "Components/ContextMenu",
  component: ContextMenu,
  tags: ["autodocs"],
  argTypes: {
    open: {
      control: {
        type: "boolean",
      },
    },
    notifyOnClose: {
      table: {
        disable: true,
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof ContextMenu>;

export const Shown: Story = {
  render: (args) => <SetupComponent {...args} />,
  args: {
    items: [
      {
        id: "item-1",
        text: "Item 1",
        icon: <CloudDownloadIcon />,
        action: action("Item 1 clicked"),
        disabled: false,
      },
      {
        id: "item-2",
        text: "Item 2",
        icon: <SendOutlined />,
        action: action("Item 2 clicked"),
        disabled: false,
      },
    ],
  },
};

export const DisabledMenuItems: Story = {
  render: (args) => <SetupComponent {...args} />,
  args: {
    items: [
      {
        id: "item-1",
        text: "Item 1",
        icon: <CloudDownloadIcon />,
        action: action("Item 1 clicked"),
        disabled: true,
      },
      {
        id: "item-2",
        text: "Item 2",
        icon: <SendOutlined />,
        action: action("Item 2 clicked"),
        disabled: true,
      },
    ],
  },
};

export const WithoutIcons: Story = {
  render: (args) => <SetupComponent {...args} />,
  args: {
    items: [
      {
        id: "item-1",
        text: "Item 1",
        action: action("Item 1 clicked"),
        disabled: false,
      },
      {
        id: "item-2",
        text: "Item 2",
        action: action("Item 2 clicked"),
        disabled: false,
      },
    ],
  },
};

export const Mixed: Story = {
  render: (args) => <SetupComponent {...args} />,
  args: {
    items: [
      {
        id: "item-1",
        text: "Item 1",
        icon: <CloudDownloadIcon />,
        action: action("Item 1 clicked"),
        disabled: false,
      },
      {
        id: "item-2",
        text: "Item 2",
        action: action("Item 2 clicked"),
        disabled: true,
      },
      {
        id: "item-3",
        text: "Item 2",
        action: action("Item 2 clicked"),
        disabled: false,
      },
      {
        id: "item-4",
        text: "Item 3",
        icon: <SendOutlined />,
        action: action("Item 3 clicked"),
        disabled: true,
      },
    ],
  },
};

function SetupComponent(props: Readonly<ContextMenuProps>) {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  useEffect(() => {
    const anchor = document.getElementById("anchor-on-me");
    setAnchorEl(anchor);
  }, []);

  return (
    <>
      <Box display="flex" justifyContent="center" alignItems="center" height="100%">
        <div style={{ border: "1px dashed", height: "20px", width: "100px" }} id="anchor-on-me" />
      </Box>
      <ContextMenu
        anchorEl={anchorEl}
        open={anchorEl !== null}
        notifyOnClose={action("notifyOnClose")}
        items={props.items}
      />
    </>
  );
}
