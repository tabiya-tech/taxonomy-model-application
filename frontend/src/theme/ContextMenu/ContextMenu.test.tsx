// mute the console
import "src/_test_utilities/consoleMock";

import { fireEvent, render, waitFor, screen, within } from "@testing-library/react";
import ContextMenu, { ContextMenuProps, DATA_TEST_ID } from "./ContextMenu";
import React from "react";
import { MenuItemConfig } from "./menuItemConfig.types";

describe("ContextMenu", () => {
  const stdGivenProps: ContextMenuProps = {
    anchorEl: document.createElement("div"), // Mock an HTMLElement
    open: true,
    notifyOnClose: jest.fn(),
    items: [],
  };

  describe("render tests", () => {
    test("renders correctly the menu open", () => {
      // GIVEN the following menu items
      const givenItems: MenuItemConfig[] = [
        // an enabled  menu item with an icon
        {
          id: "1",
          text: "foo",
          icon: <div />,
          disabled: false,
          action: jest.fn(),
        },
        // a menu item without an icon
        {
          id: "2",
          text: "bar",
          icon: undefined,
          disabled: false,
          action: jest.fn(),
        },
        // a disabled menu item
        {
          id: "3",
          text: "baz",
          icon: <div />,
          disabled: true,
          action: jest.fn(),
        },
      ];

      // WHEN the component is rendered with the given items and
      render(<ContextMenu {...stdGivenProps} items={givenItems} />);

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
      // AND the ContextMenu to be visible
      const actualMenu = screen.getByTestId(DATA_TEST_ID.MENU);
      expect(actualMenu).toBeInTheDocument();
      // AND the menu items to be visible
      const actualMenuItems = screen.getAllByTestId(DATA_TEST_ID.MENU_ITEM);
      expect(actualMenuItems).toHaveLength(givenItems.length);

      // AND every menu item
      actualMenuItems.forEach((menuItem, index) => {
        // to have the correct text in the right order
        const actualText = within(menuItem).getByTestId(DATA_TEST_ID.MENU_ITEM_TEXT);
        expect(actualText).toBeInTheDocument();
        expect(actualText.textContent).toEqual(givenItems[index].text);
      });

      // AND the menu item with the icon to have an icon
      const actualIcon = within(actualMenuItems[0]).getByTestId(DATA_TEST_ID.MENU_ITEM_ICON);
      expect(actualIcon).toBeInTheDocument();
      // AND the menu item without an icon to not have an icon
      const actualIcon2 = within(actualMenuItems[1]).queryByTestId(DATA_TEST_ID.MENU_ITEM_ICON);
      expect(actualIcon2).not.toBeInTheDocument();
      // AND the menu item with the enabled state to be enabled
      expect(actualMenuItems[0]).not.toHaveAttribute("aria-disabled", "true");
      // AND the menu item with the disabled state to be disabled
      expect(actualMenuItems[2]).toHaveAttribute("aria-disabled", "true");

      // AND to match the snapshot
      expect(actualMenu).toMatchSnapshot();
    });

    test("renders correctly the menu closed", () => {
      // GIVEN that the menu is closed
      const givenProps = { ...stdGivenProps, open: false };

      // WHEN the component is rendered
      render(<ContextMenu {...givenProps} />);

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
      // AND the ContextMenu to not be shown
      const actualMenu = screen.queryByTestId(DATA_TEST_ID.MENU);
      expect(actualMenu).not.toBeInTheDocument();
    });
  });

  describe("action tests", () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test("on click on a menu item, the notifyOnClose and then the item's action are called", async () => {
      // GIVEN some menu items
      const givenItems: MenuItemConfig[] = [
        {
          id: "1",
          text: "foo",
          icon: <div />,
          disabled: false,
          action: jest.fn(),
        },
        {
          id: "2",
          text: "bar",
          icon: <div />,
          disabled: false,
          action: jest.fn(),
        },
      ];

      // WHEN the component is rendered
      render(<ContextMenu {...stdGivenProps} items={givenItems} />);
      // AND the user clicks on the first menu item
      const actualMenuItem = screen.getAllByTestId(DATA_TEST_ID.MENU_ITEM)[0];
      fireEvent.click(actualMenuItem);

      // THEN expect the notifyOnClose to be called first
      await waitFor(() => expect(stdGivenProps.notifyOnClose).toHaveBeenCalled());
      // AND the action of the clicked  menu item to be called
      expect(givenItems[0].action).toHaveBeenCalledAfter(stdGivenProps.notifyOnClose as jest.Mock);
      // AND the action of the second menu item to not be called
      expect(givenItems[1].action).not.toHaveBeenCalled();
    });

    test("on close the notifyOnClose should be called", () => {
      // GIVEN the ContextMenu is rendered with the std given props
      render(<ContextMenu {...stdGivenProps} />);

      // WHEN the use closed the menu ( by pressing the escape key )
      const actualMenu = screen.getByTestId(DATA_TEST_ID.MENU);
      fireEvent.keyDown(actualMenu, { key: "Escape" });

      // THEN the notifyOnClose should be called to close the menu
      expect(stdGivenProps.notifyOnClose).toHaveBeenCalled();
    });
  });
});
