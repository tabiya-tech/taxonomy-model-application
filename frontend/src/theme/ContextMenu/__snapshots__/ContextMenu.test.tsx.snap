// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu render tests renders correctly the menu open 1`] = `
<div
  class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
  data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiMenu-paper MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiPopover-paper css-xem7h2-MuiPaper-root-MuiMenu-paper-MuiPopover-paper"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
    tabindex="-1"
  >
    <ul
      class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
      role="menu"
      tabindex="-1"
    >
      <li
        class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
        data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item"
        role="menuitem"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item-icon"
        >
          <span
            aria-hidden="true"
            class="material-icons notranslate MuiIcon-root MuiIcon-colorPrimary MuiIcon-fontSizeMedium css-shoean-MuiIcon-root"
          >
            <div />
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item-text"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
          >
            foo
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
        data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item-text"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
          >
            bar
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </li>
      <li
        aria-disabled="true"
        class="MuiButtonBase-root MuiMenuItem-root Mui-disabled MuiMenuItem-gutters Mui-disabled MuiMenuItem-root Mui-disabled MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
        data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item-icon"
        >
          <span
            aria-hidden="true"
            class="material-icons notranslate MuiIcon-root MuiIcon-colorPrimary MuiIcon-fontSizeMedium css-shoean-MuiIcon-root"
          >
            <div />
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          data-testid="b7499b01-8082-4209-8667-c7d559a70caf-menu-item-text"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
          >
            baz
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
